arch/cortex-m0/src/lib.rs

```diff
@@ -3,7 +3,75 @@
 
 extern crate kernel;
 
-#[no_mangle]
+pub mod nvic;
+
+#[cfg(not(target_os = "none"))]
+pub unsafe extern "C" fn generic_isr() {}
+
+#[cfg(target_os = "none")]
+#[naked]
+/// All ISRs are caught by this handler which disables the NVIC and switches to the kernel.
+pub unsafe extern "C" fn generic_isr() {
+    asm!("
+    /* Skip saving process state if not coming from user-space */
+    ldr r0, MEXC_RETURN_PSP
+    cmp lr, r0
+    bne _ggeneric_isr_no_stacking
+
+    /* We need the most recent kernel's version of r1, which points */
+    /* to the Process struct's stored registers field. The kernel's r1 */
+    /* lives in the second word of the hardware stacked registers on MSP */
+    mov r1, sp
+    ldr r1, [r1, #4]
+    str r4, [r1, #16]
+    str r5, [r1, #20]
+    str r6, [r1, #24]
+    str r7, [r1, #28]
+
+    push {r4-r7}
+    mov  r4, r8
+    mov  r5, r9
+    mov  r6, r10
+    mov  r7, r11
+    str r4, [r1, #0]
+    str r5, [r1, #4]
+    str r6, [r1, #8]
+    str r7, [r1, #12]
+    pop {r4-r7}
+
+    ldr r0, MEXC_RETURN_MSP
+_ggeneric_isr_no_stacking:
+    /* Find the ISR number by looking at the low byte of the IPSR registers */
+    mrs r0, IPSR
+    movs r1, #0xff
+    ands r0, r1
+    /* ISRs start at 16, so substract 16 to get zero-indexed */
+    subs r0, r0, #16
+
+    /* NVIC.ICER[r0 / 32] = 1 << (r0 & 31) */
+	movs	r2, #31
+	asrs	r3, r0, #31
+	ands	r3, r2
```

Comment:
Now I'm a bit more confused, as it looks like this is and'ing with the shifted r0? I think commenting each step will alleviate confusion here.